### PR TITLE
Add getAllStoredItems to lib/browser-storage

### PR DESCRIPTION
**Pull request by sgomes**
Since I'm still waiting on feedback in #37883, I decided to move forward with implementing the new functionality in `lib/browser-storage` in preparation. If it's not needed, I can drop the PR. If it is needed, then it's already done 🙂 

#### Changes proposed in this Pull Request

* Add `getAllStoredItems` to `lib/browser-storage`
* Fix some lint issues

#### Testing instructions

No testing instructions. This is new functionality that's not being used anywhere yet.

### CODE REVIEW

File: client/lib/browser-storage/index.ts
Comment (jsnajdr) [inline client/lib/browser-storage/index.ts]: This function (and also `idbGetAll`) should be able to load only keys that match a certain pattern.

In our case, if the user ID is 1234, we want to load keys that are in format
```
redux-state-1234(:[a-z]+)?
```

If the Indexed DB also contains state for user 5678, we should ignore the `redux-state-5678` keys.

There's one special case: if the Signup section cannot find its `redux-state-1234:signup` key, it should load the `redux-state-logged-out:signup` key as a fallback. That's used when signing up a new user and transferring the signup state across the logged-out/logged-in boundary. At some point during the signup, the new user becomes logged-in and the signup continues. And wants to use state stored during the logged-out part.

Comment (jsnajdr) [reply client/lib/browser-storage/index.ts]: I'm not sure if the `{ [ key: string ]: T }` type is correct. The stored values don't necessarily all have the same type `T`. They are generic objects (results of `JSON.parse`), but can theoretically also be strings or numbers.

Comment (sgomes) [reply client/lib/browser-storage/index.ts]: > This function (and also `idbGetAll`) should be able to load only keys that match a certain pattern.
> 
> In our case, if the user ID is 1234, we want to load keys that are in format
> 
> ```
> redux-state-1234(:[a-z]+)?
> ```
> 
> If the Indexed DB also contains state for user 5678, we should ignore the `redux-state-5678` keys.
> 
> There's one special case: if the Signup section cannot find its `redux-state-1234:signup` key, it should load the `redux-state-logged-out:signup` key as a fallback. That's used when signing up a new user and transferring the signup state across the logged-out/logged-in boundary. At some point during the signup, the new user becomes logged-in and the signup continues. And wants to use state stored during the logged-out part.

Thanks for the detailed explanation, @jsnajdr ! Considering what you just mentioned, it sounds like it's best to keep things simple at the `lib/browser-storage` and accept a regular expression as a pattern to match. The logic can then stay in whatever module is using `lib/browser-storage`.

Comment (sgomes) [reply client/lib/browser-storage/index.ts]: > I'm not sure if the `{ [ key: string ]: T }` type is correct. The stored values don't necessarily all have the same type `T`. They are generic objects (results of `JSON.parse`), but can theoretically also be strings or numbers.

You're right, I don't think that applies here. Created a new type, which unfortunately must contain an explicit `any` 😞

**Comment by sgomes** [comment]
@jsnajdr Thanks for the feedback, addressed all of it.

This is getting complex enough that it warrants some unit tests, I think. I'll write some.

**Comment by sgomes** [comment]
@jsnajdr: Tests added.

**Comment by sgomes** [comment]
Rebased on master to pick up node 12 changes.

**Comment by sgomes** [comment]
Rebased on master. Still waiting for a review.

**Comment by sgomes** [comment]
Added everyone else (I think) that expressed an interest in modularising Calypso state further. This PR is a requirement for moving the experiment in #37883 forward, and unlocking the rest of the work.

**Comment by sgomes** [comment]
The E2E test failures seem unrelated; as they should be, since the new functionality here isn't yet being used anywhere!

File: client/lib/browser-storage/bypass.ts
Comment (dmsnell) [inline client/lib/browser-storage/bypass.ts]: given the very generic nature of `StoredItems` I'm suspicious of its value in `types.ts`. I acknowledge that I may be pre-judging the name `StoredItems` based on an assumption that's it's generic.

we're also doing something interesting here which is decoupling the type of the actual storage from the type of the returned value. I think we have no idea what types of data we'll store here and so I can image value in either leaving the types as they were, defining `StoredItems` here in this file, making the type in `types.ts` more specific (such as naming it `JSONSerializable`), or going further to synchronize the types

```ts
// types.ts
type JSONValue =
  | string
  | number
  | boolean
  | null
  | Array<JSONValue>

type JSONSerializable = {
  [key: string]: JSONValue | JSONSerializable;
};
```

```ts
// bypass.ts
const memoryStore = new Map<string, JSONSerializable>();

export async function getAllStoredItems( pattern?: RegExp ): Promise< JSONSerializable > {
 	// TODO: Replace below with Object.fromEntries when it's available (needs core-js@3).
 	const results: JSONSerializable = {};
 	for ( const [ key, value ] of memoryStore.entries() ) {
 		if ( ! pattern || pattern?.test( key ) ) {
 			results[ key ] = value;
 		}
 	}
 	return results;
 }
```

Mostly musing here, but the use of `JSONSerializable` would prevent a number of possible cases trying to store objects with known types that cannot be serializable.

```ts
const j: JSONSerializable = {
  foo: () => { }, // Type '() => void' is not assignable to type 'string | number | boolean | JSONValue[] | JSONSerializable | null'.
  bar: new Map(), // Type 'Map<any, any>' is not assignable to type 'string | number | boolean | JSONValue[] | JSONSerializable | null'.
}
```

File: client/lib/browser-storage/index.ts
Comment (blowery) [reply client/lib/browser-storage/index.ts]: Maybe we should be creating a separate `objectStore` for each user id?

Comment (blowery) [reply client/lib/browser-storage/index.ts]: Not sure how we'd work that with the fallback-to-localStorage part though. 

To elaborate on that, I think partitioning each user into their own objectStore would allow us to cleanly clear each store without having to worry about stepping on others toes. It may also allow us to implement the bypass store in a way that more closely mimics how the normal backing store works and make that path more reliable when testing...

Comment (sgomes) [reply client/lib/browser-storage/index.ts]: I'm not opposed to using a different `objectStore` per user if it helps work around browser issue. 

It wouldn't fit well into the existing architecture, but we can always rework it, either by attempting to merge the two layers, or by adding extra functionality to this layer. As for the `localStorage` fallback, it would handle that as well, with a noop if need be.

**Comment by sgomes** [comment]
Rebased on master (and squashed, to make it a little bit easier).

File: client/lib/browser-storage/index.ts
Comment (dmsnell) [reply client/lib/browser-storage/index.ts]: > You're right, I don't think that applies here. Created a new type, which unfortunately must contain an explicit any 😞

We could create a type which carries through but it'd be useless since these values only inhabit a type at runtime.

If we want to keep the safety up then we might prefer `unknown` to `any` since that will require people to verify that the data they're using matches their expectations.

Screen Shot 2020-01-16 at 10 12 31 AM
[Playground Link](http://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=2&pc=22#)

Can also give control to the callee to work around the type checks if we say we know exactly what type we expect here.

```ts
async function getMyStoredItems<T extends JSONObject>(pattern?: RegExp): Promise<T> {
 	const results: StoredItemz = {};
 	for ( const [ key, value ] of memoryStore.entries() ) {
 		if ( ! pattern || pattern?.test( key ) ) {
 			results[ key ] = value;
 		}
 	}
 	return results as T;
}

async function mine() {
  const values = await getMyStoredItems<{ name: string }>(/test/);

  // assumes that name exists and is a string
  return values.name.length;
}
```

Comment (sgomes) [reply client/lib/browser-storage/index.ts]: Thanks, @dmsnell! `unknown` looks like a very good option. I wasn't aware of how it was different from `any`, but I've since done some research and it looks like a good fit.

**Comment by blowery** [comment]
@sgomes @jsnajdr The lint error here is interesting. eslint looks up the folder structure for the first `package.json` and uses that to determine what's a valid import.

In this case `fake-indexeddb` is in the root devDependencies, but the nearest package.json is in `client`.

**Comment by blowery** [comment]
Looks like we can teach eslint to look in both package.json for valid deps. Added in https://github.com/Automattic/wp-calypso/pull/38320/commits/6efed66b60ba1e4b70d61a9103cc066a29e51c38

**Review by blowery — APPROVED** [review]
LGTM! Thanks for addressing all the review concerns.

File: client/lib/browser-storage/bypass.ts
Comment (jsnajdr) [reply client/lib/browser-storage/bypass.ts]: IndexedDB can serialize and store much more than just the types that are allowed by JSON. It can do Maps, Sets, Blobs, Files, Dates and RegExps, ArrayBuffers and ImageData. Basically any value except functions, DOM elements and Symbols. See the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) docs for more details.

Is there a TypeScript type for all that other than `any`?

I had a quick look at the [Dexie](https://github.com/dfahlander/Dexie.js) library, a popular IDB wrapper. It's written in TypeScript and uses `any` for the object store values, too.

File: client/lib/browser-storage/index.ts
Comment (jsnajdr) [inline client/lib/browser-storage/index.ts]: Another unneeded optional chaining? You just checked if `pattern` is truthy and yet use `pattern?.test`.

File: client/lib/browser-storage/index.ts
Comment (jsnajdr) [inline client/lib/browser-storage/index.ts]: I suspect that nowadays `Object.entries( window.localStorage )` works well, too, and should even work on IE11 with the CoreJS polyfills.

**Review by jsnajdr — APPROVED** [review]
Functionality-wise this should work great, thanks! I posted a handful of suggestions how to improve the code quality.

File: client/.eslintrc.js
Comment (sgomes) [reply client/.eslintrc.js]: Thanks! It's probably best for me to wait until that PR is merged and then rebase on top of it, right?

File: client/lib/browser-storage/bypass.ts
Comment (sgomes) [reply client/lib/browser-storage/bypass.ts]: That's right! Hooray! 😄 I'll make the change here.

File: client/.eslintrc.js
Comment (jsnajdr) [reply client/.eslintrc.js]: Yes! I just merged it a few seconds ago.

Comment (sgomes) [reply client/.eslintrc.js]: Rebased! 🙂

File: client/lib/browser-storage/index.ts
Comment (sgomes) [reply client/lib/browser-storage/index.ts]: That's a very good point, thanks!

File: client/lib/browser-storage/index.ts
Comment (sgomes) [reply client/lib/browser-storage/index.ts]: Yup, much simpler, thank you!

File: client/lib/browser-storage/index.ts
Comment (sgomes) [reply client/lib/browser-storage/index.ts]: Reading through the documentation I can't seem to find any situations where the event or target would be null. I'll get rid of the optional chaining.

File: client/lib/browser-storage/index.ts
Comment (sgomes) [reply client/lib/browser-storage/index.ts]: Right, totally unnecessary.

File: client/lib/browser-storage/index.ts
Comment (sgomes) [reply client/lib/browser-storage/index.ts]: Yes, it does seem to work! Thanks!

**Comment by sgomes** [comment]
@jsnajdr : Looks like we're still getting lint errors on account of `fake-indexeddb`. Any ideas?

**Comment by blowery** [comment]
@sgomes https://github.com/Automattic/wp-calypso/pull/38320#issuecomment-575297774 :D

**Comment by sgomes** [comment]
> @sgomes [#38320 (comment)](https://github.com/Automattic/wp-calypso/pull/38320#issuecomment-575297774) :D

Right, my dropping your PRs was intentional, since I rebased on top of #38893, which appeared to do the same. Probably should have mentioned it, sorry 🙂 

Did I miss something, though? Was your PR more comprehensive, and that's why the error persists?

**Comment by blowery** [comment]
> Did I miss something, though? Was your PR more comprehensive, and that's why the error persists?

Yeah, they're subtly different, it's easy to miss. My change allows for dependencies to come from either the current package.json, or the root. The landed PR only allows for them in the current package.json, which lacks the devDependencies from the root, so the rule fails. 

Both are wrong though, so we'll take it up in another PR

**Pull request merged**
